viewing_party/party.py

```diff
@@ -1,23 +1,165 @@
+
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+
+    total = 0
+
+    if not user_data["watched"]:
+        return 0.0
+    
+    for movie in user_data["watched"]:
+        total += movie["rating"]
+    
+    average = total / len(user_data["watched"])
+
+    return average
+
+def get_most_watched_genre(user_data):
+    
+    compare_dict = {}
+
+    if not user_data["watched"]:
+        return None
+
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre not in compare_dict:
+            compare_dict[genre] = 0
+        compare_dict[genre] += 1
+    
+    most_watched_genre = max(compare_dict, key=compare_dict.get)
+
+    return most_watched_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+
+    friends_movie_list = []
+
+    for friend in user_data["friends"]:
+        friends_movie_list.extend(friend["watched"])
+    
+    unique_movies = []
+
+    for movie in user_data["watched"]:
+        if movie not in friends_movie_list:
+            unique_movies.append(movie)
+
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+
+    friends_movie_list = []
+
+    for friend in user_data["friends"]:
+        friends_movie_list.extend(friend["watched"])
+    
+    unique_movies = []
+
+    for movie in friends_movie_list:
+        if movie not in user_data["watched"] and movie not in unique_movies:
+            unique_movies.append(movie)
+
+    return unique_movies
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+
+    unique_movies = get_friends_unique_watched(user_data)
+    friends_rec = []
+
+    for movie in unique_movies:
+        if movie["host"] in user_data["subscriptions"]:
+            friends_rec.append(movie)
+
+    return friends_rec
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+
+    if "favorites" not in user_data:
+        return []
+
+    fav_genre_set = set()
+
+    for fav_movie in user_data["favorites"]:
+        fav_genre_set.add(fav_movie["genre"])
+    
+    friends_rec = get_available_recs(user_data)
+
+    recommendations = []
+
+    for movie in friends_rec:
+        if movie["genre"] in fav_genre_set:
+            recommendations.append(movie)
+
+
+    return recommendations
+
+def get_rec_from_favorites(user_data):
+
+    if "favorites" not in user_data:
+        return []
+
+    friends_movie_list = []
+
+    for friend in user_data["friends"]:
+        friends_movie_list.extend(friend["watched"])
+

```

Comment:
Rather than getting the friends list here and then later using it to filter out user movies that the friends have watched, we might assume that for a user to have a movie as a favorite, they must have watched the movie. And we have a function that gets us the movies a user has watched that the friends have not. Could this help us simplify this function?